auctions/english-auction/src/utils.sw

```diff
@@ -0,0 +1,125 @@
+library utils;
+
+dep abi;
+dep data_structures;
+dep errors;
+
+use abi::NFT;
+use data_structures::{Asset, NFTAsset, TokenAsset};
+use errors::{AccessError, InputError};
+use std::{
+    assert::require,
+    chain::auth::{AuthError, msg_sender},
+    context::{call_frames::{contract_id, msg_asset_id}, msg_amount},
+    identity::Identity,
+    option::Option,
+    result::Result,
+    token::transfer,
+};
+
+/// This function returns true if the `to` `Identity` is approved to transfer the token.
+///
+/// # Arguments
+///
+/// * `from` - The `Identity` which owns the NFTs.
+/// * `to` - The `Identity` which the NFTs should be transfered to.
+/// * `asset` - The `NFTAsset` struct which contains the NFT data.
+pub fn approved_for_nft_transfer(from: Identity, to: Identity, asset: NFTAsset) -> bool {
+    let nft_contract = asset.contract_id;
+    // TODO: This will be a Vec
+    let nft_id = asset.token_ids;
+
+    let nft_abi = abi(NFT, nft_contract.value);
+    let approved_for_all = nft_abi.is_approved_for_all(from, to);
+    // TODO: This needs to loop over a Vec of token_ids
+    let approved: Option<Identity> = nft_abi.get_approved(nft_id);
+
+    // The to address either needs to be approved for all or approved for this token id
+    approved_for_all || (approved.is_some() && to == approved.unwrap())
+}
+
+/// This function returns true if the `owner` `Identity` owns the NFT token.
+///
+/// # Arguments
+///
+/// * `owner` - The `Identity` which owns the tokens.
+/// * `asset` - The `NFTAsset` struct which contains the NFT data.
+pub fn owns_nft(owner: Identity, asset: NFTAsset) -> bool {
+    let nft_contract = asset.contract_id;
+    // TODO: This will be a Vec
+    let token_id = asset.token_ids;
+
+    let nft_abi = abi(NFT, nft_contract.value);
+    // TODO: This will need to loop over a Vec of token_ids
+    let token_owner: Option<Identity> = nft_abi.owner_of(token_id);
+    token_owner.is_some() && owner == token_owner.unwrap()
+}
+
+/// This function will return the identity of the sender.
+pub fn sender_identity() -> Identity {
+    let sender: Result<Identity, AuthError> = msg_sender();
+    sender.unwrap()
+}
+
+/// This function will transfer the `Asset` given to the specified `Identity`.
+pub fn transfer_asset(to: Identity, asset: Asset) {
+    match asset {
+        Asset::NFTAsset(asset) => {
+            transfer_nft(Identity::ContractId(contract_id()), to, asset)
```

Comment:
I think this might not work as you expect it because it is outside a contract. 

https://github.com/FuelLabs/sway/blob/c3869c960969b51eb0fc9b9b26f75260bcf287ba/sway-lib-std/src/context/call_frames.sw#L20-L27